### DISCUSSION TITLE
자료구조 - 스택/큐/덱 - 호현&일구&성은

### DISCUSSION THREAD

**Discussion by seongeun42** [Shortudy-10th/tech-for-developer#44] [질의 응답]
- 큐, 스택, 덱이 각각 어떠한 구조를 가지는지 설명해 주세요.
- 스택, 큐, 덱을 프로그램상에서 구현한다면, 어떻게 구현할 수 있을까요?
- OS, 어플리케이션 등 다양한 프로그램에서 스택, 큐, 덱을 실제로 어떻게 활용하고 있나요? 그 예시에 대해 설명해주세요.
- 브라우저에서 스택과 큐는 언제 활용되고 있나요?
- 스택과 큐의 특징과 이 둘은 어디에 주로 사용되나요?

**Comment by seongeun42** [comment]
### ❓️큐, 스택, 덱이 각각 어떠한 구조를 가지는지 설명해 주세요.

<br>

### 💬 답변
<blockquote>

- 스택(Stack)
  - LIFO(Last In First Out) : 마지막에 들어온 것이 먼저 나가는 구조
  - e.g. 짐 쌓기, 함수 호출

- 큐(Queue)
  - FIFO(First In First Out) : 먼저 들어온 것이 먼저 나가는 구조
  - e.g 줄 서기

- 덱(Deque)
  - 스택과 큐의 특징을 모두 가지고 있는 자료구조로, 앞뒤로 모두 넣고 뺄 수 있는 구조

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️스택, 큐, 덱을 프로그램 상에서 구현한다면, 어떻게 구현할 수 있을까요?

<br>

### 💬 답변
<blockquote>

* 스택: 배열, 링크드 리스트, 동적 할당 3가지 모두 구현 가능하나, 링크드 리스트를 활용해 구현하는 것이 제일 적합하다고 생각한다. 자바 내부를 뜯어보면 vector로 구현해놨다.
* 큐: FIFO 특징을 고려했을때, 배열과 링크드 리스트로 구현할 수 있다. 배열의 경우 최대 사이즈가 제한되므로 링크드 리스트가 더 적합하다고 생각한다.
* 덱: circular array를 통해 구현할 수 있다. 또는 링크드 리스트를 사용하고 리스트의 맨 앞을 가리키는 front와 맨 뒤를 가리키는 rear를 설정하여 구현할 수도 있다. 
   * 자바에서 Deque을 구현한 대표적인 클래스인 ArrayDeque를 보면 데이터를 Object[]로 관리하고 있음을 확인할 수 있다.


</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️스택만을 이용해서 큐를 구현해보세요.

<br>

### 💬 답변
<blockquote>

두 개의 스택을 이용해 구현한다. A 스택에 쌓인 데이터를 B 스택에 전부 옮기면, 역순으로 쌓이게 되며 A 스택에 제일 먼저 들어간 게 B 스택에서는 가장 위에 위치하게 된다. B 스택을 사용하면 큐와 같은 동작을 구현할 수 있다.

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️스택만을 이용해서 덱을 구현해보세요.

<br>

### 💬 답변
<blockquote>

두 개의 스택을 이용해 구현한다. Front 스택과 Back 스택으로 나누어 덱의 앞, 뒤 요청에 따라 두 개의 스택에 나눠서 처리한다. 만약 한쪽 스택이 비었다면 다른 스택의 원소를 빈 스택으로 옮겨서 처리하는 식으로 구현한다.

</blockquote>

<br>

**Reply by seongeun42** [reply]
### ❓️C++의 stack과 queue, Java의 Stack과 Queue는 각각 어떻게 구현되었는지 간략하게 설명해주세요.

<br>

### 💬 답변
<blockquote>

C++에서 stack과 queue는 내부적으로 deque를 사용해 deque의 기능 중 일부를 막아 stack과 queue에 적합한 기능만 쓸 수 있도록 하는 식으로 구현되었다. Java의 Stack 클래스는 Vector 클래스를 상속받아 구현되었고, Queue 인터페이스는 ArrayDeque나 LinkedList를 통해 구현한다.

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ OS, 어플리케이션 등 다양한 프로그램에서 스택, 큐, 덱을 실제로 어떻게 활용하고 있나요? 그 예시에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

* 스택
  * JVM에서 메소드를 호출할 때, 스택을 사용한다. 메소드가 호출되면, 스택 영역에 스택 프레임이 생성되고 그 안에 지역변수나 매개변수 값들이 저장된다. 스택의 LIFO 특징에 의해, 프로그램이 실행되면서 메소드 범위를 벗어나면 프레임이 pop되면서 자연스레 지역변수와 매개변수 값들이 제거된다. 
  * 브라우저의 뒤로가기나 실행 취소(undo) 같은 경우도 이전에 실행했던 것들을 스택 구조로 쌓아뒀다가 가장 마지막에 실행한 결과를 취소하는 식으로 처리한다.

* 큐: 자바의 스레드풀, 운영체제에서 프로세스 스케줄링, 노드에서 패킷을 교환할 때 등 다양한 상황에서 쓰인다.

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 브라우저에서 스택과 큐는 언제 활용되고 있나요?

<br>

### 💬 답변
<blockquote>

![img](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/fc94bede-4e8b-47ad-81c3-659841fc64b1)

웹은 HTML, CSS, JavaScript로 작성하고, 웹브라우저는 이 문서들을 해석해 실행시켜준다.

JavaScript 엔진은 내부적으로 메모리 힙(Memory Heap)과 콜 스택(Call Stack)을 가지고 있는데, 메모리 힙은 메모리 할당이 이루어지는 곳이고, 콜 스택은 코드 실행에 따라 함수 호출 스택이 쌓이는 곳이다. 콜 스택은 이름 그대로 스택 구조이므로, 가장 마지막에 쌓인 코드를 먼저 수행하게 된다.

단, JavaScript는 기본적으로 싱글 스레드 기반 언어이므로, 콜 스택을 하나만 가지고 있으며 한 번에 하나의 작업만 처리할 수 있다. 하지만 이로 인해 하나의 작업이 끝날 때까지 나머지 작업들이 계속 대기하면 프로그램이 매우 느려질 수 있기 때문에, 이벤트 루프와 브라우저에 존재하는 Web API 같은 백그라운드 공간, 태스크 큐 등을 통해 비동기 방식을 실현한다. 우선 바로 실행할 수 있는 코드만 먼저 수행하고 `setTimeout`과 같이 비동기로 실행되는 것들은 해당 함수의 콜백함수와 타이머가 백그라운드 공간으로 옮겨진다. 이후 타이머가 만료되면 태스크 큐에 넣어진다. 이후 콜 스택이 전부 비게 되면 태스크 큐에 들어온 순서대로 콜 스택으로 옮겨져 실행된다.

</blockquote>

<br>

**Reply by sikutisa** [reply]
잘 모르겠어서 일단 링크만 https://haenny.tistory.com/243
누가 대신 해주면 더 좋고

**Comment by seongeun42** [comment]
### ❓️ 스택과 큐의 특징과 이 둘은 어디에 주로 사용되나요?

<br>

### 💬 답변
<blockquote>



</blockquote>

<br>

**Reply by sikutisa** [reply]
이 질문 좀 중복되는 질문 같은데 어떻게 생각하나요

**Reply by seongeun42** [reply]
같은 의견이에요. 위의 질문들이랑 같은 내용인 거 같습니다.
